### PR TITLE
Added a script to recover VLANs

### LINKED ISSUE #232: Add script to find missing VLANs from interfaces
Related to kytos issue [#509](https://github.com/kytos-ng/kytos/issues/509)

### PR DESCRIPTION
**Pull request by Alopalao**
Closes #232

### Summary

Added script to make available VLANs that are not used by paths.

### Local Tests
Run  the script in a 2023.2 environment.

### End-to-End Tests
N/A

### CODE REVIEW

**Comment by Alopalao** [comment]
This script is best after solution from `mef_eline` issue [#543](https://github.com/kytos-ng/mef_eline/issues/543) is performed.

**Review by viniarck — CHANGES_REQUESTED** [review]
@Alopalao, overall the approach looks good to me. Quickly explored in dev env while similating some missing vlans. Things to fix though: 

- Console script location. This isn't a traditional `scripts/db/<version>` script, since it's meant to be copy and pasted in the kytosd console, let's add in a new folder tree like `scripts/console/<version>`? We could go for just `scripts/` but that is for other general scripts not related to console. 

- This script isn't copy paste-able. Rename it to be `.py` too, then format it with black, let's ensure it works, initially it didn't work for me:

```
  Cell In[3], line 52
    if of_lldp_vlan:
    ^
IndentationError: expected an indented block after 'for' statement on line 51
```

I'll add the rest of other review points in new comments.

File: scripts/db/2023.2.0/002_recover_vlans
Comment (viniarck) [inline scripts/db/2023.2.0/002_recover_vlans:59]: 1 - It would be very informative to also print what will be set in terms of `available_tags` for operators to know that will happen. 

2 - Speaking of which, can you also add support for a `DRY_RUN` boolean flag? Network operators should set as True if they want to see what will be made available on that interface. 

This gives the operator a last opportunity to check, if they're are OK, which we don't expected surprises then they run without DRY_RUN subsequently. This is kind of analagous to our aggregated db commands in other traditional DB cli scripts.

File: scripts/db/2023.2.0/002_recover_vlans
Comment (viniarck) [inline scripts/db/2023.2.0/002_recover_vlans:49]: What about potentially missing vlans from EVPLs UNIs? Especially since version 2023.2.0 where it's allocating, can you review and double check this part? I can see a case where a UNI wasn't allocating before (the new script would consider it available when computing the tag range diff), then after a new update it can try to release then it'd would result in error at runtime later on.

File: scripts/db/2023.2.0/README.md
Comment (viniarck) [inline scripts/db/2023.2.0/README.md:156]: Can you also include the `use_tags.py` console script and add information about it in the README? Initially it was used as quick workaround, but since that it's the corollary-ish version of this script it can go hand in hand and can be run whenever operators need to sanitize to make sure all s-vlans have been allocated including tagged UNIs:

```python
from collections import defaultdict

DRY_RUN = False

mef_eline = controller.napps[("kytos", "mef_eline")]
evcs = {
    evc_id: evc.as_dict()
    for evc_id, evc in mef_eline.circuits.items()
    if not evc.archived
}

in_use_tags = defaultdict(list)
for evc_id, evc in evcs.items():
    for link in evc["current_path"]:
        svlan = link["metadata"]["s_vlan"]["value"]
        intfa = link["endpoint_a"]["id"]
        intfb = link["endpoint_b"]["id"]
        in_use_tags[intfa].append((svlan, evc_id))
        in_use_tags[intfb].append((svlan, evc_id))
    for link in evc["failover_path"]:
        svlan = link["metadata"]["s_vlan"]["value"]
        intfa = link["endpoint_a"]["id"]
        intfb = link["endpoint_b"]["id"]
        in_use_tags[intfa].append((svlan, evc_id))
        in_use_tags[intfb].append((svlan, evc_id))

    for uni in ("uni_a", "uni_z"):
        intf_id = evc[uni]["interface_id"]
        if (
            "tag" in evc[uni]
            and "tag_type" in evc[uni]["tag"]
            and evc[uni]["tag"]["tag_type"] in ("vlan", 1)
        ):
            tag = evc[uni]["tag"]["value"]
            if isinstance(tag, int):
                in_use_tags[intf_id].append((tag, evc_id))
            elif isinstance(tag, list):
                for tag_item in tag:
                    if isinstance(tag_item, int):
                        in_use_tags[intf_id].append((tag_item, evc_id))
                    elif isinstance(tag_item, list) and len(tag_item) == 1:
                        in_use_tags[intf_id].append((tag_item[0], evc_id))
                    elif isinstance(tag_item, list) and len(tag_item) == 2:
                        for val in range(tag_item[0], tag_item[1]+1):
                            in_use_tags[intf_id].append((val, evc_id))


for intf_id in in_use_tags:
    intf = controller.get_interface_by_id(intf_id)
    for tag, evc_id in in_use_tags[intf_id]:
        if intf.is_tag_available(tag, tag_type="vlan"):
            dry_run_key = "WILL" if not DRY_RUN else "WOULD"
            print(
                f"s_vlan {tag} that was in use from EVC {evc_id} is still available on intf {intf_id}, {dry_run_key} use it..."
            )
            if not DRY_RUN:
                intf.use_tags(controller, tag)

```

**Review by viniarck — CHANGES_REQUESTED** [review]
Finally, changelog also needs to include a `General Information` information section mentioning about these console scripts. Just a quick summary since the README.md explains the rest, but we need that for future information in the release notes.

File: scripts/db/2023.2.0/002_recover_vlans
Comment (Alopalao) [reply scripts/db/2023.2.0/002_recover_vlans]: Added the option to `DRY_RUN`

File: scripts/db/2023.2.0/002_recover_vlans
Comment (Alopalao) [reply scripts/db/2023.2.0/002_recover_vlans]: Added UNIs VLANs to the collection.

File: scripts/console/2023.2.0/README.md
Comment (viniarck) [inline scripts/console/2023.2.0/README.md:7]:
```suggestion
Some paths were not dealocating VLANs from their respective interfaces. This script [`001_use_tags.py`](./001_use_tags.py) identifies them and allocates them.
```

File: scripts/console/2023.2.0/README.md
Comment (viniarck) [inline scripts/console/2023.2.0/README.md:5]:
```suggestion
<details><summary><h3>Use VLANs in <code>interface.available_vlans</code></h3></summary>
```

File: scripts/console/2023.2.0/001_use_tags.py
Comment (viniarck) [inline scripts/console/2023.2.0/001_use_tags.py:30]: I just caught an issue on this part that I tried to contribute, for EPLs in since it's reading directly from `mef_eline.circuits`, `tag` can be None too, we'll need this extra safe guard, can you apply these patches below on both console scripts?

```
     27 for uni in ("uni_a", "uni_z"):
     28     intf_id = evc[uni]["interface_id"]
     29     if (
     30         "tag" in evc[uni]
---> 31         and "tag_type" in evc[uni]["tag"]
     32         and evc[uni]["tag"]["tag_type"] in ("vlan", 1)
     33     ):
     34         tag = evc[uni]["tag"]["value"]
     35         if isinstance(tag, int):

TypeError: argument of type 'NoneType' is not iterable
```

```
diff --git a/scripts/console/2023.2.0/001_use_tags.py b/scripts/console/2023.2.0/001_use_tags.py
index 4639be9..6e8c534 100644
--- a/scripts/console/2023.2.0/001_use_tags.py
+++ b/scripts/console/2023.2.0/001_use_tags.py
@@ -28,6 +28,7 @@ for evc_id, evc in evcs.items():
         intf_id = evc[uni]["interface_id"]
         if (
             "tag" in evc[uni]
+            and evc[uni]["tag"]
             and "tag_type" in evc[uni]["tag"]
             and evc[uni]["tag"]["tag_type"] in ("vlan", 1)
         ):
diff --git a/scripts/console/2023.2.0/002_recover_vlans.py b/scripts/console/2023.2.0/002_recover_vlans.py
index db47bce..eda2072 100644
--- a/scripts/console/2023.2.0/002_recover_vlans.py
+++ b/scripts/console/2023.2.0/002_recover_vlans.py
@@ -62,6 +62,7 @@ for evc_id, evc in evcs.items():
         intf_id = evc[uni]["interface_id"]
         if (
             "tag" in evc[uni]
+            and evc[uni]["tag"]
             and "tag_type" in evc[uni]["tag"]
             and evc[uni]["tag"]["tag_type"] in ("vlan", 1)
         ):
```

**Review by viniarck — APPROVED** [review]
@Alopalao fantastic scripts and docs. 

There's one minor safe guard that needs to be adjusted check out my comment. 

I also explored it again locally, behaved as expected, awesome to also have the informative and user friendly consolidaded prints that you added:

```
Missing available tags in interface 00:00:00:00:00:00:00:01:3:
WRONG -> [[1, 3798], [3800, 4095]]
CORRECT -> [[2, 3798], [3800, 4095]]
MISSING -> []
```

I'll leave it pre-approved @Alopalao, once it gets fixed feel free to merge and then announce on Slack dev that the script is ready, and then also ask if anybody wants to still wants to try it out or explore it, otherwise let's ship it.

**Review by viniarck — APPROVED** [review]
img

**Pull request merged**
